src/rust/src/rustzcash.rs

```diff
@@ -403,23 +328,24 @@ fn priv_get_note(
     pk_d: *const [c_uchar; 32],
     value: u64,
     r: *const [c_uchar; 32],
-) -> Result<Note<Bls12>, ()> {
+) -> Result<Note, ()> {
     let diversifier = Diversifier(unsafe { *diversifier });
-    let g_d = diversifier.g_d::<Bls12>(&JUBJUB).ok_or(())?;
+    let g_d = diversifier.g_d().ok_or(())?;
 
-    let pk_d = edwards::Point::<Bls12, Unknown>::read(&(unsafe { &*pk_d })[..], &JUBJUB)
-        .map_err(|_| ())?;
+    let pk_d = de_ct(jubjub::ExtendedPoint::from_bytes(unsafe { &*pk_d })).ok_or(())?;
 
-    let pk_d = pk_d.as_prime_order(&JUBJUB).ok_or(())?;
+    let pk_d = de_ct(pk_d.into_subgroup()).ok_or(())?;
 
     // Deserialize randomness
-    let r = Fs::from_repr(read_fs(unsafe { &*r })).map_err(|_| ())?;
+    // If this is after ZIP 212, the caller has calculated rcm, and we don't need to call
+    // Note::derive_esk, so we just pretend the note was using this rcm all along.
+    let rseed = Rseed::BeforeZip212(de_ct(jubjub::Scalar::from_bytes(unsafe { &*r })).ok_or(())?);
```

Comment:
Checking my understanding: the reason that we use `()` as the error type is to avoid revealing information about the type of failure?

Reply:
In this particular case, it's used because we don't actually care to expose the errors (we don't have the ability yet to do so across FFI), so mapping all errors to `()` enables them all to be uniformly raised. In the places we call `priv_get_note`, we match the `Err(_)` case and return `false`. This could more simply be represented by having `priv_get_note` return `Option<Note>`, but `Result<Note, ()>` gives more of an indication that the failure case is an actual failure.